File: backend/app/schemas/hal_import.py
"""PROJ-21: Schemas für HAL-Import, Ergebnis-Screening und Shortlist."""

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


class HalImportRunRead(BaseModel):
    id: UUID
    total_files: int
    status_imported: int
    status_unchanged: int
    status_updated: int
    status_failed: int
    created_at: datetime


class HalImportedFileRow(BaseModel):
    id: UUID
    import_run_id: UUID
    origin_path: str
    content_hash: str
    import_version: int
    processing_status: str
    error_message: str | None = None
    created_at: datetime


class HalImportFileResult(BaseModel):
    origin_path: str
    content_hash: str
    status: str  # importiert / unverändert / aktualisiert / fehlerhaft
    error_message: str | None = None
    strategy_name: str | None = None


class HalImportResponse(BaseModel):
    import_run_id: UUID
    total: int
    files: list[HalImportFileResult]


class HalAssignRequest(BaseModel):
    strategy_version_id: UUID | None = None


class HalResultRead(BaseModel):
    id: UUID
    imported_file_id: UUID
    strategy_name: str
    asset: str
    timeframe: str
    period_start: date
    period_end: date | None = None
    net_return_pct: float
    max_drawdown_pct: float
    trade_count: int
    sortino_ratio: float | None = None
    profit_factor: float | None = None
    sharpe_ratio: float | None = None
    win_rate_pct: float | None = None
    cagr_pct: float | None = None
    calmar_ratio: float | None = None
    report_link: str | None = None
    parameters: dict | list | None = None
    long_short_breakdown: dict | None = None
    direction: str | None = None
    fee_pct: float | None = None
    slippage_ticks: float | None = None
    sizing_model: str | None = None
    strategy_version_id: UUID | None = None
    assignment_origin: str | None = None
    import_version: int
    created_at: datetime


class HalUnassignedRead(HalResultRead):
    import_origin_path: str
    suggested_version_id: UUID | None = None
    suggested_version_name: str | None = None


class ShortlistEntry(BaseModel):
    strategy_version_id: UUID
    created_at: datetime


SUCCESS_GROUP_CALMAR_MIN = 0.8
SUCCESS_GROUP_SORTINO_MIN = 0.5
SUCCESS_GROUP_MIN_TRADES_PER_YEAR = 6

File: nextjs_app/lib/schemas/hal-import.ts
import { z } from "zod";

export const halImportFileResultSchema = z.object({
  origin_path: z.string(),
  content_hash: z.string(),
  status: z.enum(["importiert", "unverändert", "aktualisiert", "fehlerhaft"]),
  error_message: z.string().nullable().optional(),
  strategy_name: z.string().nullable().optional(),
});

export const halImportResponseSchema = z.object({
  import_run_id: z.string(),
  total: z.number().int(),
  files: z.array(halImportFileResultSchema),
});

export const halImportRunRowSchema = z.object({
  id: z.string(),
  total_files: z.number().int(),
  status_imported: z.number().int(),
  status_unchanged: z.number().int(),
  status_updated: z.number().int(),
  status_failed: z.number().int(),
  created_at: z.string(),
});

export const halImportedFileRowSchema = z.object({
  id: z.string(),
  import_run_id: z.string(),
  origin_path: z.string(),
  content_hash: z.string(),
  import_version: z.number().int(),
  processing_status: z.string(),
  error_message: z.string().nullable(),
  created_at: z.string(),
});

export const halUnassignedSchema = z.object({
  id: z.string(),
  imported_file_id: z.string(),
  strategy_name: z.string(),
  asset: z.string(),
  timeframe: z.string(),
  period_start: z.string(),
  period_end: z.string().nullable(),
  net_return_pct: z.number(),
  max_drawdown_pct: z.number(),
  trade_count: z.number().int(),
  sortino_ratio: z.number().nullable(),
  profit_factor: z.number().nullable(),
  sharpe_ratio: z.number().nullable(),
  import_origin_path: z.string(),
  suggested_version_id: z.string().nullable(),
  suggested_version_name: z.string().nullable(),
  strategy_version_id: z.string().nullable(),
  created_at: z.string(),
});

export const versionSummarySchema = z.object({
  id: z.string(),
  family_id: z.string(),
  version_number: z.number().int(),
  name: z.string().nullable(),
  frozen_at: z.string(),
});

export type HalImportFileResult = z.infer<typeof halImportFileResultSchema>;
export type HalImportResponse = z.infer<typeof halImportResponseSchema>;
export type HalImportRunRow = z.infer<typeof halImportRunRowSchema>;
export type HalImportedFileRow = z.infer<typeof halImportedFileRowSchema>;
export type HalUnassigned = z.infer<typeof halUnassignedSchema>;
export type VersionSummary = z.infer<typeof versionSummarySchema>;

export const STATUS_LABELS: Record<string, string> = {
  importiert: "Importiert",
  unverändert: "Unverändert",
  aktualisiert: "Aktualisiert",
  fehlerhaft: "Fehlerhaft",
};

export const STATUS_BADGE_VARIANT: Record<string, "default" | "secondary" | "destructive" | "outline"> = {
  importiert: "default",
  unverändert: "outline",
  aktualisiert: "secondary",
  fehlerhaft: "destructive",
};
